app.py

```diff
@@ -48,6 +48,13 @@ def logout():
 
 @app.route("/")
 def index():
+    if not (app.config["CLIENT_ID"] and app.config["CLIENT_SECRET"]):
```

Comment:
Could this return templates/missing_config.hml and put the HTML in there? It's worrisome when I see HTML inside a Python file, and it's also invalid HTML since it's lacking doctype, html, head, title, body, etc?

Also, I see you're saying `not app.config["CLIENT_ID"]` versus `"CLIENT_ID" not in app.config`. Is that because it's set to an empty string at this point? If it's set to `None`, we'd need the latter check, I think.

Reply:
> Could this return templates/missing_config.hml and put the HTML in there? It's worrisome when I see HTML inside a Python file, and it's also invalid HTML since it's lacking doctype, html, head, title, body, etc?

Due to the lack of the familiar Flask Debug output, and the lack of useful Log Stream (mentioned above), I find this little check-and-hint here tremendously helpful for the *very first run*. After that, the app developer would probably just remove these two lines. For that reason, I do not want to overdo it otherwise the develop would need to - but likely forget to - delete one more file. You know, [practicality beats purity](https://zen-of-python.info/although-practicality-beats-purity.html#9).

> Also, I see you're saying not app.config["CLIENT_ID"] versus "CLIENT_ID" not in app.config. Is that because it's set to an empty string at this point? If it's set to None, we'd need the latter check, I think.

It is not about the empty string. Our sample would set it to None by default. Yet it doesn't harm to be robust to guard against a potential empty string input.

But, regardless of the None-vs-empty-string situation, `"CLIENT_ID" not in app.config` is never the semantics we want here. We do not just want an existence of a key, we need some real value to be consume.

Reply:
I was worried when I saw dict access since typically a not-found-key raises a KeyError, but I realize now that app_config will set CLIENT_ID to None if its not found, so CLIENT_ID will always be a key. So this logic is correct, sorry for false alarm.

Reply:
> I was worried when I saw dict access since typically a not-found-key raises a KeyError

Understood. To guard against that KeyError yet _still_ check the value (rather than just checking the key), the proper way is to `if dictionary.get("KEY")` rather than `if "KEY" in dictionary`.

> false alarm

Don't mention it. I'm grateful to have an extra pair of eyes reviewing. Better safe than sorry. :-)

Reply:
lol, funny timing, I actually taught dict.get() in a Python class yesterday-
https://pamelafox.github.io/python-crash-course/week3.html#/11/0/3

Reply:
Nice! Whom do you teach those classes to? Your course material is neat, and much more polished than [my old school attempt](https://github.com/rayluo/first-programming-book/wiki). May I borrow your materials when I and some other volunteers teach our next class?

Reply:
Oh nice, an e book! I teach for corise.com, but you're free to use the content.

Lecture slides at:
https://pamelafox.github.io/python-crash-course/

Exported articles/exercises at:
https://pamelafox.github.io/python-crash-course/content/en/
(Doesn't have all the rich widgets of the corise.com platform, but the text is there.)